Gems/Terrain/Code/Source/Components/TerrainPhysicsColliderComponent.cpp

```diff
@@ -350,42 +343,37 @@ namespace Terrain
+            ([[maybe_unused]] size_t xIndex, [[maybe_unused]] size_t yIndex, const AzFramework::SurfaceData::SurfacePoint& surfacePoint, bool terrainExists)
         {
-            const float y = row * gridResolution.GetY() + worldSize.GetMin().GetY();
-            for (int32_t col = 0; col < gridWidth; col++)
+            float height = surfacePoint.m_position.GetZ();
+
+            // Any heights that fall outside the range of our bounding box will get turned into holes.
+            if ((height < worldHeightBoundsMin) || (height > worldHeightBoundsMax))
             {
-                const float x = col * gridResolution.GetX() + worldSize.GetMin().GetX();
-                float height = 0.0f;
-
-                bool terrainExists = true;
-                AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
-                    height, &AzFramework::Terrain::TerrainDataRequests::GetHeightFromFloats, x, y,
-                    AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT, &terrainExists);
-
-                // Any heights that fall outside the range of our bounding box will get turned into holes.
-                if ((height < worldHeightBoundsMin) || (height > worldHeightBoundsMax))
-                {
-                    height = worldHeightBoundsMin;
-                    terrainExists = false;
-                }
-
-                // Find the best surface tag at this point.
-                AzFramework::SurfaceData::SurfaceTagWeight surfaceWeight;
-                AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
-                    surfaceWeight, &AzFramework::Terrain::TerrainDataRequests::GetMaxSurfaceWeightFromFloats, x, y,
-                    AzFramework::Terrain::TerrainDataRequests::Sampler::DEFAULT, nullptr);
-
-                Physics::HeightMaterialPoint point;
-                point.m_height = height - worldCenterZ;
-                point.m_quadMeshType = terrainExists ? Physics::QuadMeshType::SubdivideUpperLeftToBottomRight : Physics::QuadMeshType::Hole;
-
-                Physics::MaterialId materialId = FindMaterialIdForSurfaceTag(surfaceWeight.m_surfaceType);
-                point.m_materialIndex = GetMaterialIdIndex(materialId, materialList);
-
-                heightMaterials.emplace_back(point);
+                height = worldHeightBoundsMin;
+                terrainExists = false;
             }
-        }
+
+            // Find the best surface tag at this point.
+            // We want the MaxSurfaceWeight. The ProcessSurfacePoints callback has surface weights sorted.
+            // So, we pick the value at the front of the list.
+            AzFramework::SurfaceData::SurfaceTagWeight surfaceWeight;
+            if (!surfacePoint.m_surfaceTags.empty())
+            {
+                surfaceWeight = *surfacePoint.m_surfaceTags.begin();
+            }
```

Comment:
The looping implementation was calling GetHeight() and GetMaxSurfaceWeights(). I'm instead using the ProcessSurfacePoint here which contains a list of ordered surface weights. We take the top value which is what the GetMaxSurfaceWeight implementation does. The down side is that we also compute the normals for the SurfacePoint which we don't need here. This would be good to revisit after a pass of optimizing the ProcessRegion functions to see if it might be better to call the ProcessHeightsFromRegion + ProcessSurfaceWeightsFromRegion. Or maybe add a flag parameter to GetSurfacePoint which could indicate what combination of Heights/Normals/SurfaceWeights we need?

Reply:
One of the earlier versions of the API proposal had a flag that could get passed in.  I took it out for the RFC because it seemed unnecessarily complex at the time - I was worried it would make the API too clunky to use especially if we had other APIs for getting the various specific bits (height, normal, weights).  But we can definitely discuss & revisit!